src/core/library.c

```diff
@@ -685,31 +708,83 @@ MsQuicLibraryUninitialize(
 
     MsQuicLibraryFreePartitions();
 
+    MsQuicLib.LazyInitComplete = FALSE;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+MsQuicLibraryUninitialize(
+    void
+    )
+{
+    QuicTraceEvent(
+        LibraryUninitialized,
+        "[ lib] Uninitialized");
+
+    MsQuicLibraryLazyUninitialize();
+
+    if (MsQuicLib.Storage != NULL) {
+        CxPlatStorageClose(MsQuicLib.Storage);
+        MsQuicLib.Storage = NULL;
+    }
+
     QuicSettingsCleanup(&MsQuicLib.Settings);
 
     CXPLAT_FREE(MsQuicLib.DefaultCompatibilityList, QUIC_POOL_DEFAULT_COMPAT_VER_LIST);
     MsQuicLib.DefaultCompatibilityList = NULL;
 
     CxPlatDispatchRwLockUninitialize(&MsQuicLib.StatelessRetry.Lock);
 
+    if (MsQuicLib.RegistrationCloseCleanupWorker) {
+        MsQuicLib.RegistrationCloseCleanupShutdown = TRUE;
+        CxPlatEventSet(MsQuicLib.RegistrationCloseCleanupEvent);
+        CxPlatThreadWait(&MsQuicLib.RegistrationCloseCleanupWorker);
+        CxPlatThreadDelete(&MsQuicLib.RegistrationCloseCleanupWorker);
+        MsQuicLib.RegistrationCloseCleanupWorker = 0;
+    }
+
+    CxPlatEventUninitialize(MsQuicLib.RegistrationCloseCleanupEvent);
+    CxPlatLockUninitialize(&MsQuicLib.RegistrationCloseCleanupLock);
+
     if (MsQuicLib.ExecutionConfig != NULL) {
         CXPLAT_FREE(MsQuicLib.ExecutionConfig, QUIC_POOL_EXECUTION_CONFIG);
         MsQuicLib.ExecutionConfig = NULL;
     }
 
-    MsQuicLib.LazyInitComplete = FALSE;
-
-    QuicTraceEvent(
-        LibraryUninitialized,
-        "[ lib] Uninitialized");
-
 #ifndef _KERNEL_MODE
     CxPlatWorkerPoolDelete(MsQuicLib.WorkerPool);
     MsQuicLib.WorkerPool = NULL;
 #endif
     CxPlatUninitialize();
 }
 
+CXPLAT_THREAD_CALLBACK(RegistrationCleanupWorker, Context)
+{
+    UNREFERENCED_PARAMETER(Context);
+
+    while (!MsQuicLib.RegistrationCloseCleanupShutdown) {
+        CxPlatEventWaitForever(MsQuicLib.RegistrationCloseCleanupEvent);
```

Comment:
I think this needs to check `RegistrationCloseCleanupShutdown` before attempting to take the lock because an APC could cause a spurious wake of this thread and in that case it should go back to waiting on the event.

Reply:
A non-alertable wait (QUIC should not be using alertable waits?) can be interrupted by kernel APCs but the wait is not abandoned/control does not return to the caller.